src/core/common/trickle_timer.cpp

```diff
@@ -53,6 +53,61 @@ TrickleTimer::TrickleTimer(Instance &aInstance, Handler aHandler)
 {
 }
 
+void TrickleTimer::GetLastTimerStart(TimeMilli &aStartTime)
+{
+    aStartTime = TimerMilli::GetFireTime();
+
+    switch (mPhase)
+    {
+    case kBeforeRandomTime:
+        aStartTime -= mTimeInInterval;
+        break;
+
+    case kAfterRandomTime:
+        aStartTime -= mInterval - mTimeInInterval;
+        break;
+    }
+}
+
+void TrickleTimer::SetIntervalMax(uint32_t aIntervalMax)
+{
+    OT_ASSERT(aIntervalMax >= mIntervalMin);
+
+    VerifyOrExit(aIntervalMax != mIntervalMax);
+
+    mIntervalMax = aIntervalMax;
+
+    if (aIntervalMax < mInterval)
```

Comment:
Should we avoid setting a new fire time if the existing (randomly chosen) timer will fire within the new interval time?

One contrived scenario is slowly calling `SetIntervalMax()` repeatedly with smaller `aIntervalMax` values can starve timer fires. Thoughts?

Reply:
I'm just trying to make sure that I'm understanding the concern correctly. Is the question about whether we should just remove this condition and always re-calculate a new timer value?

Right now, the logic I'm trying to implement is that if the new `aIntervalMax` is smaller than our running timer, we will adjust it to fire at the original start time + `aIntervalMax` (emulating our last random selection picking the `aIntervalMax` value). If the running timer is less than `aIntervalMax` we just let it run to its original duration as the original duration will still meet the bounds criteria.

My goal is demonstrated with few examples below (assume a start time of 0 for all):

| `mInterval` | `aIntervalMax` | current time |  Action | 
|--|--|--|--|
| 10000 | 15000 | 100 | timer fires at 10000 |
| 15000 | 15000 | 100 | timer fires at 15000 | 
| 20000 | 15000 | 100 | timer fires at 15000 |
| 20000 | 5000 | 10000 | timer fires at 10000 |

Looking at the code for the scheduler that `TimerMilli` uses leads me to believe that a timer with a historical fire time would be placed first in the queue of the scheduler and would fire immediately next time the scheduler checks for items which need to run.

If I've implemented this correctly, and my logic seems reasonable, the result of the contrived example using `SetIntervalMax` with a smaller `aIntervalMax` (which is always greater than `mInterval`) over time is that it will fire the timer at the original time and then on the next cycle it will continue with the trickle mechanism of attempting to increase the time, but will use a new maximum value.

In the case that `aIntervalMax` is less than `mInterval` it would trigger an immediate fire of the timer and the continue on with the trickle algorithm with the `mIntervalMax` having the value `aIntervalMax`

Reply:
Ah, yes, thanks for the additional clarification. I agree always starting from the same interval start time will result in the behavior I was looking for. Thanks!